design-system/scripts/generate-design-tokens.js

```diff
@@ -21,14 +21,99 @@ const endProgram = (message) => {
   process.exit(1);
 };
 
-const TOKEN_REGEX =
-  /^(\w+(?:-\w+)(?:-\w+)?)(?:-for-(\w+(?:-\w+)?))?(?:-when-([\w-]+?))?(?:-on-([\w-]+?))?$/i;
+const ALLOWED_CSS_VALUES_IN_CHOICES = /px|none|hsla/;
+
+const isAllowedCssChoice = (choice) =>
+  choice.match(ALLOWED_CSS_VALUES_IN_CHOICES) !== null;
 
 const supportedStates = Object.keys(definitions.states);
 const supportedComponentGroups = Object.keys(definitions.componentGroups);
+const supportedVariants = Object.keys(definitions.variants || {});
 
 const designTokens = {};
 
+/*
+  Allowed patterns with examples:
+    - <attribute>-for-<component-group>
+      + background-color-for-tag
+    - <attribute>-for-<component-group>-when-<state>-
+      + background-color-for-button-when-disabled
+    - <attribute>-for-<component-group>-as-<state>-
+      + border-radius-for-button-as-big
+    - <attribute>-for-<component-group>-as-<state>-when-<state>
+      + border-for-button-as-secondary-when-hovered
+    - <attribute>-for-<component-group>-as-<state>-as-<state>-
+      + border-radius-for-button-as-icon-as-small
+    - <attribute>-for-<component-group>-as-<state>-as-<state>-when-<state>-
+      + border-radius-for-button-as-icon-as-small-when-disabled
+*/
+function parseToken(token) {
+  const parts = token.split('-');
+  let partType = 'cssRule';
+  let newVariant = false;
+  return parts.reduce(
+    (agg, part) => {
+      if (['for', 'as', 'when'].includes(part)) {
+        partType = part;
+        if (part === 'as') newVariant = true;
+        return agg;
+      }
+
+      if (partType === 'for') {
+        agg.componentGroup = `${agg.componentGroup}${
+          !!agg.componentGroup ? '-' : ''
+        }${part}`;
+      } else if (partType === 'when') {
+        agg.state = `${agg.state}${!!agg.state ? '-' : ''}${part}`;
+      } else if (partType === 'as') {
+        if (newVariant) {
+          agg.variants.push(part);
+          newVariant = false;
+        } else {
+          const lastIndex = agg.variants.length - 1;
+          agg.variants[lastIndex] = `${agg.variants[lastIndex]}-${part}`;
+        }
+      } else {
+        agg.cssRule = `${agg.cssRule}${!!agg.cssRule ? '-' : ''}${part}`;
+      }
+
+      return agg;
+    },
+    {
+      cssRule: '',
+      componentGroup: '',
+      variants: [],
+      state: '',
+    }
+  );
+}
+
+/*
+  We make sure the order of the token parts is
+    1. component group
+    2. variants (might have several of this)
+    3. state
+
+  Eg: <attribute>-for-<component-group>-as-<state>-when-<state>
+*/
+function isValidTokenName(tokenName, tokenParts) {
```

Comment:
A function to check the syntax of tokens names. Basically we want to enforce the order of the different parts.